// vue3-baidu-map-gl/es/components/overlay/circle/index.vue.js
import { defineComponent, watch, provide, renderSlot } from 'vue';
import useParentComponentEffect from '../../../hooks/useParentComponentEffect';
import { callWhenDifferentValue, warn, bindEvents } from '../../../utils';

var script = /*#__PURE__*/ defineComponent(Object.assign({
    name: 'BCircle'
}, { __name: 'index', props: {
        center: { type: null, required: true },
        radius: { type: Number, required: true },
        strokeColor: { type: String, required: false, default: '#000' },
        strokeOpacity: { type: Number, required: false, default: 1 },
        fillColor: { type: String, required: false, default: '#fff' },
        fillOpacity: { type: Number, required: false, default: 0.3 },
        strokeWeight: { type: Number, required: false, default: 2 },
        strokeStyle: { type: null, required: false, default: 'solid' },
        enableMassClear: { type: Boolean, required: false, default: true },
        enableEditing: { type: Boolean, required: false, default: false },
        enableClicking: { type: Boolean, required: false, default: true },
        geodesic: { type: Boolean, required: false, default: false },
        clip: { type: Boolean, required: false, default: true },
        visible: { type: Boolean, required: false, default: true },
        onClick: { type: null, required: false },
        onDblclick: { type: null, required: false },
        onMousedown: { type: null, required: false },
        onMouseup: { type: null, required: false },
        onMouseout: { type: null, required: false },
        onMouseover: { type: null, required: false },
        onRemove: { type: null, required: false },
        onLineupdate: { type: null, required: false }
    }, emits: [
        'initd',
        'unload',
        'click',
        'dblclick',
        'mousedown',
        'mouseup',
        'mouseout',
        'mouseover',
        'remove',
        'lineupdate'
    ], setup(__props, { emit: vueEmits }) {
        const props = __props;
        let circle;
        const { ready } = useParentComponentEffect((map) => {
            const cal = () => {
                map.removeOverlay(circle);
            };
            const init = () => {
                if (process.env.NODE_ENV !== 'production') {
                    if (!props.center)
                        return warn('BCircle', 'center props is required');
                    if (!props.radius)
                        return warn('BCircle', 'radius props is required');
                }
                const { center, radius, strokeColor, strokeOpacity, fillColor, fillOpacity, strokeWeight, strokeStyle, enableMassClear, enableEditing, enableClicking, geodesic, clip, visible } = props;
                const centerPoint = new BMapGL.Point(center.lng, center.lat);
                circle = new BMapGL.Circle(centerPoint, radius, {
                    strokeColor,
                    strokeWeight,
                    strokeOpacity,
                    strokeStyle,
                    enableMassClear,
                    enableEditing,
                    enableClicking,
                    geodesic,
                    clip,
                    fillOpacity,
                    fillColor
                });
                visible && map.addOverlay(circle);
                bindEvents(props, vueEmits, circle);
                ready(map, circle);
            };
            init();
            // 监听值变化
            watch(() => props.center, callWhenDifferentValue(setCenter), { deep: true });
            watch(() => props.radius, setRadius);
            watch(() => props.strokeColor, setStrokeColor);
            watch(() => props.strokeOpacity, setStrokeOpacity);
            watch(() => props.fillColor, setFillColor);
            watch(() => props.fillOpacity, setFillOpacity);
            watch(() => props.strokeWeight, setStrokeWeight);
            watch(() => props.strokeStyle, setStrokeStyle);
            watch(() => props.enableMassClear, setMassClear);
            watch(() => props.enableEditing, setEditing);
            watch(() => props.visible, (n) => {
                map[n ? 'addOverlay' : 'removeOverlay'](circle);
            });
            return cal;
        });
        provide('getOverlayInstance', () => circle);
        function setRadius(radius) {
            circle.setRadius(radius);
        }
        function setCenter(center) {
            if (circle) {
                circle.setCenter(new BMapGL.Point(center.lng, center.lat));
            }
        }
        function setStrokeColor(color) {
            circle.setStrokeColor(color);
        }
        function setFillColor(color) {
            circle.setFillColor(color);
        }
        function setStrokeOpacity(opacity) {
            circle.setStrokeOpacity(opacity);
        }
        function setFillOpacity(opacity) {
            circle.setFillOpacity(opacity);
        }
        function setStrokeWeight(weight) {
            circle.setStrokeWeight(weight);
        }
        function setStrokeStyle(style) {
            circle.setStrokeStyle(style);
        }
        function setMassClear(enableMassClear) {
            enableMassClear ? circle.enableMassClear() : circle.disableMassClear();
        }
        function setEditing(enableEditing) {
            enableEditing ? circle.enableEditing() : circle.disableEditing();
        }
        return (_ctx, _cache) => {
            return renderSlot(_ctx.$slots, "default");
        };
    } }));

script.__file = "packages/components/overlay/circle/index.vue";

export { script as default };
