// vue3-baidu-map-gl/es/components/control/scale/index.vue.js
import { defineComponent, watch } from 'vue';
import useParentComponentEffect from '../../../hooks/useParentComponentEffect';

var script = /*#__PURE__*/ defineComponent(Object.assign({
    name: 'BScale'
}, { __name: 'index', props: {
        anchor: { type: null, required: false, default: 'BMAP_ANCHOR_BOTTOM_LEFT' },
        offset: { type: Object, required: false, default: () => ({ x: 83, y: 18 }) },
        unit: { type: null, required: false, default: 'BMAP_UNIT_METRIC' },
        visible: { type: Boolean, required: false, default: true }
    }, emits: ['initd', 'unload'], setup(__props) {
        const props = __props;
        let scaleCtrl;
        const { ready } = useParentComponentEffect((map) => {
            const { visible, offset, anchor } = props;
            scaleCtrl = new BMapGL.ScaleControl({
                offset: new BMapGL.Size(offset.x, offset.y),
                anchor: window[anchor]
            });
            visible && map.addControl(scaleCtrl);
            setUnit();
            ready(map, scaleCtrl);
            watch(() => props.visible, (n) => {
                map[n ? 'addControl' : 'removeControl'](scaleCtrl);
            });
            return () => map.removeControl(scaleCtrl);
        });
        // 监听比例尺单位变化
        watch(() => props.unit, setUnit);
        /**
         * 设置比例尺单位制
         */
        function setUnit() {
            scaleCtrl.setUnit(window[props.unit]);
        }
        return (_ctx, _cache) => {
            return null;
        };
    } }));

script.__file = "packages/components/control/scale/index.vue";

export { script as default };
