// vue3-baidu-map-gl/es/components/overlay/polyline/index.vue.js
import { defineComponent, watch, provide, renderSlot } from 'vue';
import useParentComponentEffect from '../../../hooks/useParentComponentEffect';
import { callWhenDifferentValue, warn, pathPointsToMapPoints, bindEvents } from '../../../utils';

var script = /*#__PURE__*/ defineComponent(Object.assign({
    name: 'BPolyline'
}, { __name: 'index', props: {
        path: { type: Array, required: true },
        strokeColor: { type: String, required: false, default: '#000' },
        strokeWeight: { type: Number, required: false, default: 2 },
        strokeOpacity: { type: Number, required: false, default: 1 },
        strokeStyle: { type: null, required: false, default: 'solid' },
        enableMassClear: { type: Boolean, required: false, default: true },
        enableEditing: { type: Boolean, required: false, default: false },
        enableClicking: { type: Boolean, required: false, default: true },
        geodesic: { type: Boolean, required: false, default: false },
        clip: { type: Boolean, required: false, default: true },
        linkRight: { type: Boolean, required: false, default: true },
        visible: { type: Boolean, required: false, default: true },
        onClick: { type: null, required: false },
        onDblclick: { type: null, required: false },
        onMousedown: { type: null, required: false },
        onMouseup: { type: null, required: false },
        onMouseout: { type: null, required: false },
        onMouseover: { type: null, required: false },
        onRemove: { type: null, required: false },
        onLineupdate: { type: null, required: false }
    }, emits: [
        'initd',
        'unload',
        'click',
        'dblclick',
        'mousedown',
        'mouseup',
        'mouseout',
        'mouseover',
        'remove',
        'lineupdate'
    ], setup(__props, { emit: vueEmits }) {
        const props = __props;
        let polyline;
        const { ready } = useParentComponentEffect((map) => {
            const clear = () => {
                polyline && map.removeOverlay(polyline);
            };
            const init = () => {
                if (!props.path || !props.path.length)
                    return process.env.NODE_ENV !== 'production' && warn('BPolyline', 'path is required and not empty array');
                const { path, strokeColor, strokeWeight, strokeOpacity, strokeStyle, enableMassClear, enableEditing, enableClicking, geodesic, clip, linkRight, visible } = props;
                const pathPoints = pathPointsToMapPoints(path);
                polyline = new BMapGL.Polyline(pathPoints, {
                    strokeColor,
                    strokeWeight,
                    strokeOpacity,
                    strokeStyle,
                    enableMassClear,
                    enableEditing,
                    enableClicking,
                    geodesic,
                    linkRight,
                    clip
                });
                visible && map.addOverlay(polyline);
                bindEvents(props, vueEmits, polyline);
                ready(map, polyline);
                watch(() => props.strokeColor, setStrokeColor);
                watch(() => props.strokeOpacity, setStrokeOpacity);
                watch(() => props.strokeWeight, setStrokeWeight);
                watch(() => props.strokeStyle, setStrokeStyle);
                watch(() => props.enableMassClear, setMassClear);
                watch(() => props.enableEditing, setEditing);
            };
            init();
            // 监听值变化
            watch(() => props.path, callWhenDifferentValue((n) => {
                if (polyline) {
                    n.length ? setPath(n) : clear();
                }
                else {
                    init();
                }
            }), {
                deep: true
            });
            watch(() => props.visible, (n) => {
                map[n ? 'addOverlay' : 'removeOverlay'](polyline);
            });
            return clear;
        });
        provide('getOverlayInstance', () => polyline);
        function setPath(path) {
            polyline && polyline.setPath(pathPointsToMapPoints(path));
        }
        function setStrokeColor(color) {
            polyline && polyline.setStrokeColor(color);
        }
        function setStrokeOpacity(opacity) {
            polyline && polyline.setStrokeOpacity(opacity);
        }
        function setStrokeWeight(weight) {
            polyline && polyline.setStrokeWeight(weight);
        }
        function setStrokeStyle(style) {
            polyline && polyline.setStrokeStyle(style);
        }
        function setMassClear(enableMassClear) {
            if (polyline) {
                enableMassClear ? polyline.enableMassClear() : polyline.disableMassClear();
            }
        }
        function setEditing(enableEditing) {
            if (polyline) {
                enableEditing ? polyline.enableEditing() : polyline.disableEditing();
            }
        }
        return (_ctx, _cache) => {
            return renderSlot(_ctx.$slots, "default");
        };
    } }));

script.__file = "packages/components/overlay/polyline/index.vue";

export { script as default };
